File: backend/apiquanlynhahang/apiquanlynhahang/Controllers/MangVeController.cs
using apiquanlynhahang.BLL.Services;
using apiquanlynhahang.Common;
using apiquanlynhahang.DTOs.Requests;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace apiquanlynhahang.Controllers;

[ApiController]
[Route("api/mang-ve/don-hang")]
[Authorize]
public class MangVeController : ControllerBase
{
    private readonly DonHangService _donHangService;
    private readonly AuthService _authService;

    public MangVeController(DonHangService donHangService, AuthService authService)
    {
        _donHangService = donHangService;
        _authService = authService;
    }

    [HttpPost]
    public async Task<IActionResult> TaoDonMangVe([FromBody] TaoDonMangVeDto dto, CancellationToken cancellationToken)
    {
        var current = User.LayNguoiDungHienTai();
        if (current is null) return Unauthorized(new { message = "Khong xac dinh duoc nguoi dung hien tai" });

        var thongTinNguoiDung = await _authService.LayThongTinNguoiDungHienTaiAsync(current.MaND, cancellationToken);
        if (thongTinNguoiDung is null || string.IsNullOrWhiteSpace(thongTinNguoiDung.MaKH))
        {
            return BadRequest(new { message = "Khong tim thay ma khach hang de tao don mang ve" });
        }

        dto.MaKH = thongTinNguoiDung.MaKH;
        var duLieu = await _donHangService.TaoDonMangVeAsync(dto, cancellationToken);
        return StatusCode(201, new { data = duLieu });
    }

    [HttpGet("{maDonHang}")]
    public async Task<IActionResult> LayDonMangVe(string maDonHang, CancellationToken cancellationToken)
    {
        var current = User.LayNguoiDungHienTai();
        if (current is null) return Unauthorized(new { message = "Khong xac dinh duoc nguoi dung hien tai" });

        var thongTinNguoiDung = await _authService.LayThongTinNguoiDungHienTaiAsync(current.MaND, cancellationToken);
        var duLieu = await _donHangService.LayChiTietDonMangVeAsync(maDonHang, thongTinNguoiDung?.MaKH, cancellationToken);
        return duLieu is null ? NotFound(new { message = "Khong tim thay don mang ve" }) : Ok(new { data = duLieu });
    }

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpGet("/api/mang-ve/admin/don-hang")]
    public async Task<IActionResult> LayDanhSachDonMangVeChoAdmin(CancellationToken cancellationToken)
        => Ok(new { data = await _donHangService.LayDanhSachDonMangVeChoAdminAsync(cancellationToken) });

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpPatch("/api/mang-ve/admin/don-hang/{maDonHang}/trang-thai")]
    public async Task<IActionResult> CapNhatTrangThaiDonMangVe(string maDonHang, [FromBody] CapNhatTrangThaiDto dto, CancellationToken cancellationToken)
    {
        var duLieu = await _donHangService.CapNhatTrangThaiDonMangVeAsync(maDonHang, dto.TrangThai, cancellationToken);
        return duLieu is null ? NotFound(new { message = "Khong tim thay don mang ve" }) : Ok(new { data = duLieu });
    }
}

File: backend/nest-api/src/modules/mang-ve/mang-ve.controller.ts
import { Body, Controller, Get, Headers, Param, Patch, Post } from '@nestjs/common';
import { MangVeService } from './mang-ve.service';

type BanGhi = Record<string, any>;

@Controller('api/mang-ve')
export class MangVeController {
  constructor(private readonly mangVeService: MangVeService) {}

  @Post('don-hang')
  taoDonMangVe(@Headers('authorization') authorization: string | undefined, @Body() body: BanGhi) {
    return this.mangVeService.taoDonMangVe(authorization, body);
  }

  @Get('don-hang/:maDonHang')
  layDonMangVe(@Headers('authorization') authorization: string | undefined, @Param('maDonHang') maDonHang: string) {
    return this.mangVeService.layDonMangVe(authorization, maDonHang);
  }

  @Get('noi-bo/don-hang')
  layDanhSachDonMangVeChoNoiBo(@Headers('authorization') authorization?: string) {
    return this.mangVeService.layDanhSachDonMangVeChoNoiBo(authorization);
  }

  @Patch('noi-bo/don-hang/:maDonHang/trang-thai')
  capNhatTrangThaiDonMangVe(@Headers('authorization') authorization: string | undefined, @Param('maDonHang') maDonHang: string, @Body() body: BanGhi) {
    return this.mangVeService.capNhatTrangThaiDonMangVe(authorization, maDonHang, String(body.trangThai || ''));
  }

  @Get('lich-su')
  layLichSuDonMangVe(@Headers('authorization') authorization?: string) {
    return this.mangVeService.layLichSuDonMangVe(authorization);
  }

  @Patch('don-hang/:maDonHang/huy')
  huyDonMangVe(@Headers('authorization') authorization: string | undefined, @Param('maDonHang') maDonHang: string) {
    return this.mangVeService.huyDonMangVe(authorization, maDonHang);
  }
}
